corehq/motech/generic_inbound/middleware/hl7.py

```diff
@@ -0,0 +1,159 @@
+
+from corehq.motech.generic_inbound.exceptions import GenericInboundUserError
+from corehq.motech.generic_inbound.middleware.base import BaseApiMiddleware
+from corehq.motech.generic_inbound.utils import ApiResponse
+
+
+@attr.s(kw_only=True, frozen=True, auto_attribs=True)
+class Hl7ApiResponse(ApiResponse):
+    hl7_response: str
+
+    def _get_http_response(self):
+        return HttpResponse(status=self.status, content=self.hl7_response)
+
+
+class Hl7Middleware(BaseApiMiddleware):
+    """API middleware for handling HL7 v2 payloads"""
+
+    def __init__(self, request_data):
+        super().__init__(request_data)
+        self.hl7_message = None
+
+    def _get_body_for_eval_context(self):
+        try:
+            self.hl7_message = parse_hl7(self.request_data.data)
+            msh = self.hl7_message.msh
+            return {
+                "version": msh.version_id.value,
+                "message_type": {
+                    "code": msh.message_type.message_code.value,
+                    "event": msh.message_type.trigger_event.value,
+                    "structure": msh.message_type.message_structure.value,
+                },
+                "message": hl7_message_to_dict(self.hl7_message, False)
+            }
+        except HL7apyException as e:
+            raise GenericInboundUserError(gettext("Error parsing HL7: {}").format(str(e)))
+
+    def get_success_response(self, response_json):
+        hl7_response = self._get_ack("success", 200)
+        return Hl7ApiResponse(status=200, internal_response=response_json, hl7_response=hl7_response)
+
+    def _get_generic_error(self, status_code, message):
+        hl7_response = self._get_ack(message, status_code)
+        return Hl7ApiResponse(status=status_code, internal_response={'error': message}, hl7_response=hl7_response)
+
+    def _get_submission_error_response(self, status_code, form_id, message):
+        hl7_response = self._get_ack(message, status_code)
+        return Hl7ApiResponse(status=status_code, internal_response={
+            'error': message,
+            'form_id': form_id,
+        }, hl7_response=hl7_response)
+
+    def _get_validation_error(self, status_code, message, errors):
+        hl7_response = self._get_ack(message, status_code, errors)
+        return Hl7ApiResponse(status=status_code, internal_response={
+            'error': message,
+            'errors': errors,
+        }, hl7_response=hl7_response)
+
+    def _get_ack(self, ack_text, status_code, errors=None):
+        ack = Message("ACK", version=self.hl7_message.msh.version_id.value)
+        ack.msh.msh_5 = self.hl7_message.msh.msh_3  # receiving application
+        ack.msh.msh_6 = self.hl7_message.msh.msh_4  # receiving facility
+        ack.msh.msh_4 = self.hl7_message.msh.msh_6  # sending facility
+        ack.msh.msh_11 = self.hl7_message.msh.msh_11  # processing ID
+        ack.msh.msh_10 = self.request_data.request_id  # outbound control ID
```

Comment:
I see this is set up to be a consistent ID even when requests are reprocessed - that seems logical, though I'm curious what the impetus is.  Like, what would putting `uuid4().hex` here break?

Reply:
The main goal here was to provide some kind of auditability for the responses so that you could look up the request log based on the ID in the response.